File: backend-spring-boot/src/test/java/no/vicx/backend/user/service/UserImageServiceTest.java
package no.vicx.backend.user.service;

import no.vicx.database.user.UserImageRepository;
import no.vicx.database.user.UserRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.MediaType;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.stream.Stream;

import static no.vicx.backend.user.UserTestUtils.createMultipartFile;
import static no.vicx.backend.user.UserTestUtils.createValidVicxUser;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class UserImageServiceTest {

    @Mock
    UserService userService;

    @Mock
    UserRepository userRepository;

    @Mock
    UserImageRepository userImageRepository;

    @InjectMocks
    UserImageService sut;

    @Test
    void addOrReplaceUserImage_givenExistingUser_expectImageAddedToUser() throws IOException {
        var userInTest = spy(createValidVicxUser());
        var fileInTest = spy(createMultipartFile("test-png.png", MediaType.IMAGE_PNG_VALUE));

        when(userService.getUserByUserName(anyString())).thenReturn(userInTest);

        sut.addOrReplaceUserImage(fileInTest, userInTest.getUsername());

        verify(fileInTest).getBytes();
        verify(fileInTest).getContentType();

        verify(userInTest).setUserImage(any());

        verify(userService).getUserByUserName(userInTest.getUsername());
        verify(userRepository).save(any());
    }

    @Test
    void addOrReplaceUserImage_givenGetBytesThrowsIOException_expectIOException() throws IOException {
        var fileInTest = mock(MultipartFile.class);

        when(fileInTest.getBytes()).thenThrow(IOException.class);

        assertThrows(IOException.class, () -> sut.addOrReplaceUserImage(fileInTest, "~username~"));

        verify(userService, never()).getUserByUserName(anyString());
    }

    private static Stream<Arguments> invalidFileSource() {
        return Stream.of(
                Arguments.of(new byte[]{1, 2, 3}, null),
                Arguments.of(null, MediaType.IMAGE_PNG_VALUE)
        );
    }

    @ParameterizedTest
    @MethodSource("invalidFileSource")
    void addOrReplaceUserImage_givenInvalidFile_expectNullPointerException(
            byte[] bytes, String mediaType) throws IOException {

        var fileInTest = mock(MultipartFile.class);
        when(fileInTest.getBytes()).thenReturn(bytes);
        when(fileInTest.getContentType()).thenReturn(mediaType);

        assertThrows(NullPointerException.class, () -> sut.addOrReplaceUserImage(fileInTest, "~username~"));

        verify(userService, never()).getUserByUserName(anyString());
    }

    @Test
    void addOrReplaceUserImage_givenNonExistingUser_expectNullPointerException() {
        assertThrows(NullPointerException.class, () -> sut.addOrReplaceUserImage(
                createMultipartFile("test-jpg.jpg", MediaType.IMAGE_JPEG_VALUE),
                "~username~"));

        verify(userService).getUserByUserName(anyString());
    }

    @Test
    void deleteUserImage() {
        sut.deleteUserImage("user1");
        verify(userImageRepository).deleteByUserUsername("user1");
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/user/service/UserImageServiceTest.kt
package no.vicx.backend.user.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.spyk
import io.mockk.verify
import io.mockk.verifyOrder
import no.vicx.backend.error.NotFoundException
import no.vicx.backend.user.UserTestUtils.createMockMultipartFile
import no.vicx.backend.user.UserTestUtils.createValidVicxUser
import no.vicx.database.user.UserImageRepository
import no.vicx.database.user.UserRepository
import no.vicx.database.user.VicxUser
import org.springframework.http.MediaType

class UserImageServiceTest :
    BehaviorSpec({
        val userService: UserService = mockk()
        val userRepository: UserRepository = mockk()
        val userImageRepository: UserImageRepository = mockk(relaxed = true)

        val sut = UserImageService(userService, userRepository, userImageRepository)

        beforeContainer {
            clearAllMocks()
        }

        Given("addOrReplaceUserImage, existing user") {
            val savedUserSlot = slot<VicxUser>()
            val userInTest = spyk(createValidVicxUser())
            val fileInTest =
                spyk(
                    createMockMultipartFile(
                        "test-png.png",
                        MediaType.IMAGE_PNG_VALUE,
                    ),
                )

            beforeContainer {
                every { userService.getUserByUserName(any()) } returns userInTest

                every { userRepository.save(capture(savedUserSlot)) } answers {
                    savedUserSlot.captured.apply { id = 42L }
                }
            }

            When("calling addOrReplaceUserImage") {
                sut.addOrReplaceUserImage(fileInTest, userInTest.username)

                Then("expect image to be added to user") {
                    savedUserSlot.captured.userImage shouldNotBe null

                    verifyOrder {
                        userService.getUserByUserName(any())
                        fileInTest.bytes
                        fileInTest.contentType
                        userRepository.save(userInTest)
                    }
                }
            }
        }

        Given("addOrReplaceUserImage, non-existing user") {
            val usernameInTest = createValidVicxUser().username

            beforeContainer {
                every {
                    userService.getUserByUserName(any())
                } throws NotFoundException("User $usernameInTest not found")
            }

            When("calling addOrReplaceUserImage") {
                val thrown =
                    shouldThrow<NotFoundException> {
                        sut.addOrReplaceUserImage(
                            file = createMockMultipartFile("test-png.png", MediaType.IMAGE_PNG_VALUE),
                            username = usernameInTest,
                        )
                    }

                Then("exception message should be as expected") {
                    thrown.message shouldBe "User $usernameInTest not found"

                    verify(exactly = 0) { userRepository.save(any()) }
                }
            }
        }

        Given("deleteUserImage") {
            When("calling deleteUserImage") {
                sut.deleteUserImage("user1")

                Then("expect image to be deleted") {
                    verify { userImageRepository.deleteByUserUsername(any()) }
                }
            }
        }
    })
